### DISCUSSION TITLE
Multi-value progress bar

### DISCUSSION THREAD

**Discussion by troypoulter** [shadcn-ui/ui#3464] [Ideas]
Adding the ability for the `<Progress />` to accept multiple values and display them, mock-up below.

![image](https://github.com/shadcn-ui/ui/assets/19419349/41413e05-b179-474d-bd37-8059ba45a184)

The API for it could look something like the below, where we pass an array in to allow _n_ number of `values`.

`<Progress values={[33, 55]} />`

The earlier numbers in the array should be placed on top, with the following values placed underneath, perhaps with `zIndex`?

**Comment by lechnerio** [comment]
Hi @troypoulter, 
I don't know if it's still relevant to you, but I had a simular plan on using the progress bar and came up with this solution here, maybe this works for you as well: 

In addition I added the option to pass a color class. 

**Passing multiple values:** 
```
<Progress segments={[
   { value: 10 },
   { value: 50, color: "bg-green-500" },
   { value: 80, color: "bg-red-500" },
]} />

```

**Passing a single value:** 
```
<Progress segments={[
   { value: 50 }
]} />
```

**Result:**

![image](https://github.com/shadcn-ui/ui/assets/591317/9a0928c3-59cf-455c-8c93-2369a4c4f192)


this is the progress component:

```
import * as React from "react"
import * as ProgressPrimitive from "@radix-ui/react-progress"

import { cn } from "@/lib/utils"

type ProgressSegment = {
  value: number
  color?: string
}

type Props = React.ComponentPropsWithoutRef<typeof ProgressPrimitive.Root> & {
  segments: ProgressSegment[]
}

const Progress = React.forwardRef<
  React.ElementRef<typeof ProgressPrimitive.Root>,
  Props
>(
  ({ className, segments, ...props }, ref) => {
    const sortedSegments = segments.sort((a, b) => a.value - b.value)

    return (
      <ProgressPrimitive.Root
        ref={ref}
        className={cn(
          "relative h-6 w-full overflow-hidden rounded bg-secondary",
          className
        )}
        {...props}
      >
        {sortedSegments.map((segment, index) => (
          <ProgressPrimitive.Indicator
            key={index}
            className={cn("h-full transition-all absolute", segment.color ? segment.color : "bg-primary")}
            style={{ width: `${segment.value}%`, zIndex: sortedSegments.length - index }}
          />
        ))}
      </ProgressPrimitive.Root>
    )
  }
)

Progress.displayName = "Progress"

export { Progress }

```

**Reply by janmohn** [reply]
this is awesome, is there a way to add labels?

**Reply by lechnerio** [reply]
Haha glad I was able to help. I will look into it next week when I am back online and back from vacation ✌️

**Reply by janmohn** [reply]
Thank you! 

I did try using this component today - unfortunately the bars are not loading as they should do you know why? the data im using is correct and is being used to set the widths of the bars correctly, the colours are also being passed to the class names however the bars dont seem to be rendering - i have attached some screenshots incase you are able to identify what's wrong.
<img width="1030" alt="Screenshot 2024-07-15 at 22 02 49" src="https://github.com/user-attachments/assets/2311f6a5-37d5-45d6-9978-e59fe33538a2">
<img width="415" alt="Screenshot 2024-07-15 at 22 04 42" src="https://github.com/user-attachments/assets/f90e1cb2-60ff-4fe4-933c-3768b4119d90">

The images show the bars with the correct data - and even when i set the bar color thats supposed to have a width of 93% to black its still not showing up!

**Reply by Iamknownasfesal** [reply]
Hey @janmohn, i fixed the code. Since they didn't include going to left, since only adding width is not the solution.

Here is the code, feel free to use:

```ts
import * as React from "react";
import * as ProgressPrimitive from "@radix-ui/react-progress";

import { cn } from "@/lib/utils";

type ProgressSegment = {
  value: number;
  color?: string;
};

type Props = React.ComponentPropsWithoutRef<typeof ProgressPrimitive.Root> & {
  segments: ProgressSegment[];
};

const Progress = React.forwardRef<
  React.ElementRef<typeof ProgressPrimitive.Root>,
  Props
>(({ className, segments, ...props }, ref) => {
  const sortedSegments = segments.sort((a, b) => a.value - b.value);

  return (
    <ProgressPrimitive.Root
      ref={ref}
      className={cn(
        "relative h-6 w-full overflow-hidden rounded bg-secondary",
        className
      )}
      {...props}
    >
      {sortedSegments.map((segment, index) => (
        <ProgressPrimitive.Indicator
          key={index}
          className={cn(
            "h-full transition-all absolute",
            segment.color ? segment.color : "bg-primary"
          )}
          style={{
            width: `${segment.value}%`,
            zIndex: sortedSegments.length - index,
            left: `${sortedSegments
              .slice(0, index)
              .reduce((acc, segment) => acc + segment.value, 0)}%`,
          }}
        />
      ))}
    </ProgressPrimitive.Root>
  );
});

Progress.displayName = "Progress";

export { Progress };
```

**Reply by nandramihnea** [reply]
Late reply for anyone who wants to display multiple progress bars into a single progress bar all starting from '0px' left.
This means that having something like `segments={[{ value: 80 }, { value: 20, color: 'bg-green-200' }]}` will not translate in a 100% progress bar but only 80%.

![image](https://github.com/user-attachments/assets/6a9a1bd2-3385-41e9-9ffe-4fc7e8e37769)

```typescript
import * as React from 'react';
import * as ProgressPrimitive from '@radix-ui/react-progress';

import { cn } from '@/lib/utils';

type ProgressSegment = {
  value: number;
  color?: string;
};

type Props = React.ComponentPropsWithoutRef<typeof ProgressPrimitive.Root> & {
  segments: ProgressSegment[];
};

const Progress = React.forwardRef<
  React.ElementRef<typeof ProgressPrimitive.Root>,
  Props
>(({ className, segments, ...props }, ref) => {
  const sortedSegments = segments.sort((a, b) => a.value - b.value);

  return (
    <ProgressPrimitive.Root
      ref={ref}
      className={cn(
        'relative h-2 w-full overflow-hidden rounded bg-slate-200',
        className
      )}
      {...props}
    >
      {sortedSegments.map((segment, index) => (
        <ProgressPrimitive.Indicator
          key={index}
          className={cn(
            'h-full transition-all absolute top-0 rounded-r',
            segment.color ? segment.color : 'bg-primary'
          )}
          style={{
            width: `${segment.value}%`,
            left: '0%',
            zIndex: sortedSegments.length - index,
          }}
        />
      ))}
    </ProgressPrimitive.Root>
  );
});

Progress.displayName = 'Progress';

export { Progress };

```

**Comment by SukkaW** [comment]
Please vote up my feature request: https://github.com/radix-ui/primitives/discussions/3741 , vote up, give a thumbs up, and add your comment for your usage.

Since shadcn ui uses Radix UI, it is important for Radix UI to implement this first.
